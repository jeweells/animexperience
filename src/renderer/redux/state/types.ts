import { FStatus, Optional, TopView } from '@shared/types'
import {
  AnimeInfo,
  DeepAnimeIdSearchResult,
  EpisodeInfo,
  FollowedAnime,
  RecentAnimeInfo,
  RecommendationInfo,
  WatchHistoryItem
} from '@shared/types'
import { FollowedAnimeWStatus } from '../reducers/followedAnimes'
import { PreAllowWatch } from '../reducers/invokedLink'
import { RecentAnimeData } from '../../src/hooks/useRecentAnimes'
import { VideoOption } from '../../src/components/VideoPlayer'

export type AnimeSearchState = {
  // Boolean id
  searching?: string
  result?: Optional<DeepAnimeIdSearchResult>
  status: Partial<{
    result: FStatus
    moreResults: FStatus
  }>
}

export type FollowedAnimesState = {
  followedDict: Record<string, FollowedAnime>
  followed: FollowedAnimeWStatus[]
  status: Partial<{
    followed: FStatus
  }>
}

export type InvokedLinkState = {
  open: {
    watch?: string
  }
  preAllow: Partial<{
    watch: PreAllowWatch
  }>
  status: {
    watch?: FStatus
  }
}

export type NotificationsState = {
  message: string | null
  key: string
}

export type PeekState = {
  name?: string
  status: Partial<{
    info: FStatus
  }>
  info?: Optional<AnimeInfo>
  peeking?: string
}

export type PlayerState = {
  open?: string
  freezed?: boolean
}

export type OptionInfo = {
  name: string
  prefer?: boolean
  score: number
}
// Define a type for the slice state
export type PlayerOptionsState = {
  options?: OptionInfo[]
  history?: string[]
  preferred?: string[]
  status: Partial<{
    options: FStatus
  }>
}

export type RecommendationsState = {
  [name: string]: Partial<{
    status: FStatus
    recommendations: RecommendationInfo[]
  }>
}

export type TopviewState = {
  views: TopView[]
}

export type WatchState = {
  watching?: Optional<RecentAnimeData>
  availableVideos?: Optional<VideoOption[]>
  status: {
    availableVideos?: FStatus
    info?: FStatus
    watchEpisode?: FStatus
  }
  info?: Optional<AnimeInfo>
  showNextEpisodeButton?: boolean
  nextEpisodeTimeout: number
  autoFullScreen?: boolean
}

export type WatchedState = {
  episodes: {
    [name: string]: EpisodeInfo
  }
  recently?: RecentAnimeInfo[]
  status: {
    episodes?: FStatus
    recently?: FStatus
  }
}

export type WatchHistoryState = {
  sorted?: WatchHistoryItem[]
  status: {
    sorted?: FStatus
  }
}
