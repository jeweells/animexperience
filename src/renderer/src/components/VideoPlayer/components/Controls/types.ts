export type ControlsContext = {
  video: HTMLVideoElement | null
}
