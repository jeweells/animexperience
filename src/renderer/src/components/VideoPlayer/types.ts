import { RecentAnimeData } from '../../hooks/useRecentAnimes'
import { Optional } from '@shared/types'
import { VideoOption } from './index'

export type $IframeContents = JQuery<HTMLIFrameElement | Text | Comment | Document>
export type BasicVideoInfo = {
  option?: Optional<VideoOption>
  anime?: Optional<RecentAnimeData>
}

export type KnownOption =
  | 'mega'
  | 'fembed'
  | 'okru'
  | 'mixdrop'
  | 'streamtape'
  | 'stape'
  | 'maru'
  | 'sw'
  | 'yourupload'
