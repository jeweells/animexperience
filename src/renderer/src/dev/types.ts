import { RawDevMessage } from '@shared/types'
import { Moment } from 'moment'

export type DevMessage = RawDevMessage & {
  id: string
  at: Moment
}
