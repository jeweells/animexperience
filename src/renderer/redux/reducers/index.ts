export { animeSearch } from './animeSearch'
export { followedAnimes } from './followedAnimes'
export { invokedLink } from './invokedLink'
export { notifications } from './notifications'
export { peek } from './peek'
export { player } from './player'
export { playerOptions } from './playerOptions'
export { recommendations } from './recommendations'
export { topView } from './topView'
export { watch } from './watch'
export { watched } from './watched'
export { watchHistory } from './watchHistory'
