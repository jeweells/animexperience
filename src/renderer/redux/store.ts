import { configureStore } from '@reduxjs/toolkit'
import animeSearch from './reducers/animeSearch'
import followedAnimes from './reducers/followedAnimes'
import peek from './reducers/peek'
import player from './reducers/player'
import playerOptions from './reducers/playerOptions'
import recommendations from './reducers/recommendations'
import topView from './reducers/topView'
import watch from './reducers/watch'
import watched from './reducers/watched'
import watchHistory from './reducers/watchHistory'
import invokedLink from './reducers/invokedLink'
import notifications from './reducers/notifications'

export default configureStore({
  reducer: {
    watch,
    player,
    watched,
    recommendations,
    watchHistory,
    peek,
    topView,
    playerOptions,
    animeSearch,
    followedAnimes,
    invokedLink,
    notifications
  }
})
