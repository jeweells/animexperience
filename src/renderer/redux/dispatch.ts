import { AppDispatch } from '~/redux/utils'
import store from '~/redux/store'

export const dispatch: AppDispatch = (action) => store.dispatch(action)
