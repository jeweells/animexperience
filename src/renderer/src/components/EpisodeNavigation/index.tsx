import * as React from 'react'
import MenuRoundedIcon from '@mui/icons-material/MenuRounded'
import { peek } from '@reducers/peek'
import { player } from '@reducers/player'
import { watch } from '@reducers/watch'

import { useAppDispatch, useAppSelector } from '~/redux/utils'
import { TopView } from '@shared/types'
import IconButton from '@mui/material/IconButton'
import KeyboardArrowLeftRoundedIcon from '@mui/icons-material/KeyboardArrowLeftRounded'
import KeyboardArrowRightRoundedIcon from '@mui/icons-material/KeyboardArrowRightRounded'
import Tooltip from '@mui/material/Tooltip'
import { EPISODE_NAVIGATION } from '@selectors'

export const EpisodeNavigation: React.FC = React.memo(() => {
  const data = useAppSelector((d) => d.watch.info)
  const watching = useAppSelector((d) => d.watch.watching)
  const { min, max } = data?.episodesRange ?? {}
  const hasMin =
    typeof min === 'number' && typeof watching?.episode === 'number' && watching.episode > min
  const hasMax =
    typeof max === 'number' && typeof watching?.episode === 'number' && watching.episode < max

  const dispatch = useAppDispatch()
  const topView = useAppSelector((d) => d.topView.views[0])
  React.useLayoutEffect(() => {
    // Avoids the player to keep playing but does not close the modal
    dispatch(player.freeze(topView !== TopView.PLAYER))
  }, [topView])

  return (
    <React.Fragment>
      <Tooltip title={'Episodio anterior'} arrow>
        <IconButton
          data-testid={EPISODE_NAVIGATION.PREVIOUS_EPISODE_BUTTON}
          disabled={!hasMin}
          onClick={() => dispatch(watch.previousEpisode())}
        >
          <KeyboardArrowLeftRoundedIcon />
        </IconButton>
      </Tooltip>
      <Tooltip title={'Ver anime'} arrow>
        <IconButton
          data-testid={EPISODE_NAVIGATION.ANIME_DETAILS_BUTTON}
          onClick={() => {
            if (!watching?.name) return
            dispatch(peek.peek(watching.name))
          }}
        >
          <MenuRoundedIcon />
        </IconButton>
      </Tooltip>
      <Tooltip title={'Siguiente episodio'} arrow>
        <IconButton
          data-testid={EPISODE_NAVIGATION.NEXT_EPISODE_BUTTON}
          disabled={!hasMax}
          onClick={() => dispatch(watch.nextEpisode())}
        >
          <KeyboardArrowRightRoundedIcon />
        </IconButton>
      </Tooltip>
    </React.Fragment>
  )
})

EpisodeNavigation.displayName = 'EpisodeNavigation'

export default EpisodeNavigation
