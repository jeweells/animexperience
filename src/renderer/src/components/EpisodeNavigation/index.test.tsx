import EpisodeNavigation, { EpisodeNavigationProps } from './index'
import TopLayout from '../../../plugins/gatsby-plugin-top-layout/TopLayout'
import { MockStoreEnhanced } from 'redux-mock-store'
import { RootState } from '~/redux/state'
import { DeepPartial } from 'redux'
import { Provider } from 'react-redux'
import { TopView } from '@shared/types'
import { EPISODE_NAVIGATION } from '@selectors'

describe('EpisodeNavigation', () => {
  let props: EpisodeNavigationProps
  let initialState: DeepPartial<RootState>
  let store: MockStoreEnhanced

  const getComponent = (override: Partial<EpisodeNavigationProps> = {}) => {
    store = mockStore(initialState)
    return (
      <TopLayout>
        <Provider store={store}>
          <EpisodeNavigation {...props} {...override} />
        </Provider>
      </TopLayout>
    )
  }

  beforeEach(() => {
    props = {}

    initialState = {
      watch: {
        watching: {
          name: 'here goes title',
          date: '01/01/2022',
          episode: 5,
          img: 'here goes image',
          link: 'here goes episode link (5)'
        },
        info: {
          description: 'here goes description',
          link: 'here goes link',
          emitted: {
            from: 69,
            to: 129
          },
          related: [
            { name: 'Related name', link: 'related link', type: 'Ova' },
            { name: 'Related name', link: 'related link', type: 'Serie' },
            { name: 'Related name', link: 'related link', type: 'Película' },
            { name: 'Related name', link: 'related link', type: 'Especial' }
          ],
          episodeLink: 'here goes episode link (here goes episode replace)',
          type: 'Serie',
          image: 'here goes image',
          episodeReplace: 'here goes episode replace',
          episodesRange: {
            max: 10,
            min: 1
          },
          otherTitles: ['title1', 'title2'],
          status: 'En emisión',
          tags: ['tag1', 'tag2'],
          thumbnail: 'here goes thumbnail',
          title: 'here goes title'
        }
      },
      topView: {
        views: [TopView.PLAYER]
      }
    }
  })

  it('renders default', () => {
    const wrapper = render(getComponent())
    expect(wrapper.baseElement).toMatchSnapshot()
  })

  it('freezes player when using another modal', () => {
    initialState.topView!.views = [TopView.SEARCH]
    render(getComponent())
    expect(store.getActions()).toContainEqual(
      expect.objectContaining({
        payload: true,
        type: 'player/freeze'
      })
    )
  })

  it("doesn't freeze player when using player modal", () => {
    render(getComponent())
    expect(store.getActions()).toContainEqual(
      expect.objectContaining({
        payload: false,
        type: 'player/freeze'
      })
    )
  })

  it('calls previousEpisode when clicking left arrow', () => {
    const wrapper = render(getComponent())
    fireEvent.click(wrapper.getByTestId(EPISODE_NAVIGATION.PREVIOUS_EPISODE_BUTTON))
    expect(store.getActions()).toContainEqual(
      expect.objectContaining({
        type: 'watch/previousEpisode/pending'
      })
    )
  })

  it('calls nextEpisode when clicking right arrow', () => {
    const wrapper = render(getComponent())
    fireEvent.click(wrapper.getByTestId(EPISODE_NAVIGATION.NEXT_EPISODE_BUTTON))
    expect(store.getActions()).toContainEqual(
      expect.objectContaining({
        type: 'watch/nextEpisode/pending'
      })
    )
  })

  it('calls peeking anime when clicking menu icon', () => {
    const wrapper = render(getComponent())
    fireEvent.click(wrapper.getByTestId(EPISODE_NAVIGATION.ANIME_DETAILS_BUTTON))
    expect(store.getActions()).toContainEqual(
      expect.objectContaining({
        meta: expect.objectContaining({
          arg: initialState.watch!.watching!.name
        }),
        type: 'peek/peek/pending'
      })
    )
  })
})
