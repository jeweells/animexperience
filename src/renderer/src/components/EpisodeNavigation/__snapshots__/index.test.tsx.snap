// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpisodeNavigation renders default 1`] = `
.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: fill 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
  font-size: 1.5rem;
  font-size: inherit;
}

.c3 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-transform: unset;
}

.c0::-moz-focus-inner {
  border-style: none;
}

.c0.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

.c1 {
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: #fff;
  -webkit-transition: background-color 150ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4,0,0.2,1) 0ms;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 36px;
  min-width: 36px;
  color: #e9ebf0;
  padding: 8px 6px;
  background-color: #292d33;
  margin-bottom: 0;
  font-weight: normal;
  text-align: center;
  vertical-align: middle;
  outline: 0 !important;
  white-space: nowrap;
  border: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-size: 1.2rem;
  line-height: 1.42857143;
  border-radius: 6px;
}

.c1:hover {
  background-color: rgba(255,255,255,0.08);
}

.c1.Mui-disabled {
  background-color: transparent;
  color: rgba(255,255,255,0.3);
}

.c1:hover {
  background-color: #3c3f43;
}

.c1.Mui-disabled {
  color: rgb(255 255 255 / 25%);
}

@media print {
  .c0 {
    color-adjust: exact;
  }
}

@media (hover:none) {
  .c1:hover {
    background-color: transparent;
  }
}

<body>
  <div>
    <button
      aria-label="Episodio anterior"
      class="c0 MuiButtonBase-root c1 MuiIconButton-root MuiIconButton-sizeMedium"
      data-mui-internal-clone-element="true"
      data-testid="episode-navigation-previous-episode-button"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="c2 MuiSvgIcon-root MuiSvgIcon-fontSizeMedium"
        data-testid="KeyboardArrowLeftRoundedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M14.71 15.88 10.83 12l3.88-3.88c.39-.39.39-1.02 0-1.41a.9959.9959 0 0 0-1.41 0L8.71 11.3c-.39.39-.39 1.02 0 1.41l4.59 4.59c.39.39 1.02.39 1.41 0 .38-.39.39-1.03 0-1.42z"
        />
      </svg>
      <span
        class="c3 MuiTouchRipple-root"
      />
    </button>
    <button
      aria-label="Ver anime"
      class="c0 MuiButtonBase-root c1 MuiIconButton-root MuiIconButton-sizeMedium"
      data-mui-internal-clone-element="true"
      data-testid="episode-navigation-anime-details-button"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="c2 MuiSvgIcon-root MuiSvgIcon-fontSizeMedium"
        data-testid="MenuRoundedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M4 18h16c.55 0 1-.45 1-1s-.45-1-1-1H4c-.55 0-1 .45-1 1s.45 1 1 1zm0-5h16c.55 0 1-.45 1-1s-.45-1-1-1H4c-.55 0-1 .45-1 1s.45 1 1 1zM3 7c0 .55.45 1 1 1h16c.55 0 1-.45 1-1s-.45-1-1-1H4c-.55 0-1 .45-1 1z"
        />
      </svg>
      <span
        class="c3 MuiTouchRipple-root"
      />
    </button>
    <button
      aria-label="Siguiente episodio"
      class="c0 MuiButtonBase-root c1 MuiIconButton-root MuiIconButton-sizeMedium"
      data-mui-internal-clone-element="true"
      data-testid="episode-navigation-next-episode-button"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="c2 MuiSvgIcon-root MuiSvgIcon-fontSizeMedium"
        data-testid="KeyboardArrowRightRoundedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M9.29 15.88 13.17 12 9.29 8.12a.9959.9959 0 0 1 0-1.41c.39-.39 1.02-.39 1.41 0l4.59 4.59c.39.39.39 1.02 0 1.41L10.7 17.3c-.39.39-1.02.39-1.41 0-.38-.39-.39-1.03 0-1.42z"
        />
      </svg>
      <span
        class="c3 MuiTouchRipple-root"
      />
    </button>
  </div>
</body>
`;
